core/dbt/parser/schemas.py

```diff
@@ -914,6 +920,75 @@ def parse_patch(self, block: TargetBlock[NodeTarget], refs: ParserRef) -> None:
                 self.patch_node_config(node, patch)
 
             node.patch(patch)
+            self.validate_constraints(node)
+
+    def validate_constraints(self, patched_node):
+        error_messages = []
+        if (
+            patched_node.resource_type == "model"
+            and patched_node.config.constraints_enabled is True
+        ):
+            validators = [
+                self.constraints_schema_validator(patched_node),
+                self.constraints_materialization_validator(patched_node),
+                self.constraints_language_validator(patched_node),
+                self.constraints_data_type_validator(patched_node),
+            ]
+            error_messages = [validator for validator in validators if validator != "None"]
+
+        if error_messages:
+            original_file_path = patched_node.original_file_path
+            raise ParsingError(
+                f"Original File Path: ({original_file_path})\nConstraints must be defined in a `yml` schema configuration file like `schema.yml`.\nOnly the SQL table materialization is supported for constraints. \n`data_type` values must be defined for all columns and NOT be null or blank.{self.convert_errors_to_string(error_messages)}"
```

Comment:
> Only the SQL table materialization is supported for constraints.

Good on you for having the explicit callout here!

This does not work for:
- Views, because databases don't allow data type specifications or constraints when creating views. (They only allow column names, and those aren't actually cross-checked, just used for implicit renaming.)
- Incremental models - unless, of course, they have configured `'on_schema_change: 'fail'`! Should we disallow the more capable mechanisms for handling schema changes, though? This feels like a pretty big limitation.

*Unless!*:
1. We add a check for column types _after_ a view or temp table has been created. (For views: This could be within-transaction and before the alter-rename-swap on "older-school" databases, or after a view has already been created on "newer-school" databases that support `create or replace`.) That feels like a pretty big limitation, though - a big part of the appeal of "constraints," as opposed to `dbt test`, is that they should be checked _before_ the object has been created/replaced, or before expensive incremental processing has happened.
2. We add a check for column names + types _before_ the view / temp table has been created, by means of [`get_columns_in_query`](https://github.com/dbt-labs/dbt-core/blob/43e3fc22c4eae4d3d901faba05e33c40f1f1dc5a/core/dbt/include/global_project/macros/adapters/columns.sql#L24-L31), a.k.a. `where false limit 0`. That's a "free" query on most databases (including BQ!), there's just a bit of risk in extracting metadata + doing our own data type comparisons, rather than pushing it all down to the database.

I lean toward option 2, as a way to expand this support beyond just tables. In a future where we support project-level governance rules ("all public models must have constraints enabled"), without that support, we'd also be requiring that every public model be `materialized: table` — and that feels like a pretty big limitation. I've already talked to a handful of folks who have grown quite fond of a view-driven "consumption layer," and feel it offers them an additional lever of control between the underlying data (table) and its downstream users.

Reply:
For Option 2, I'm aligned on approach to get `incremental` models working. Did you want this implemented in this PR or another?

Reply:
Happy to have this in a follow-up issue/PR! I don't think we need it for the initial cut / enough to unlock a beta experience

Reply:
_[For follow-up issues, out of scope for this PR]_

After a bit more investigation, we'll need to use a modified version of `get_columns_in_query`, that pulls data types off the database cursor/API, if we want to also enforce that data types match before a table/view is created/replaced.